src/Streamly/Internal/Data/Stream/Ahead.hs

```diff
@@ -239,10 +240,21 @@ processHeap q heap st sv winfo entry sno stopping = loopHeap sno entry
                 -- transferring available results from heap to outputQueue.
                 void $ liftIO $ send sv (ChildYield a)
                 nextHeap seqNo
-            AheadEntryStream r ->
+            AheadEntryStream (Right r) ->
                 if stopping
                 then stopIfNeeded ent seqNo r
-                else runStreamWithYieldLimit True seqNo r
+                -- Use mrun here instead of the fold
+                else do
+                    res <- liftIO $ mrun (runStreamWithYieldLimit True seqNo r)
+                    restoreM res
+            AheadEntryStream (Left ((RunInIO runin), r)) ->
+                if stopping
+                then stopIfNeeded ent seqNo r
+                -- Use mrun here instead of the fold
+                else do
+                    res <- liftIO $ runin (runStreamWithYieldLimit True seqNo r)
+                    restoreM res
```

Comment:
Instead of using `runin`, `restoreM` here we can use it inside `runStreamWithYieldLimit`. If we remove the `Either` type these two cases (Right/Left) will just become the second case.

Reply:
I take that back, `runStreamWithYieldLimit` is recursive, so its better done here.